lib/pbench/test/unit/server/test_user_management_cli.py

```diff
@@ -58,169 +62,155 @@ def register_user():
-                    TestUserManagement.USER_TEXT,
-                    TestUserManagement.PSWD_SWITCH,
-                    TestUserManagement.PSWD_TEXT,
-                    TestUserManagement.EMAIL_SWITCH,
-                    TestUserManagement.EMAIL_TEXT,
-                    TestUserManagement.FIRST_NAME_SWITCH,
-                    TestUserManagement.FIRST_NAME_TEXT,
-                    TestUserManagement.LAST_NAME_SWITCH,
-                    TestUserManagement.LAST_NAME_TEXT,
-                    TestUserManagement.ROLE_SWITCH,
-                    "ADMN",
-                ],
-            )
-            assert result.exit_code == 2
-            assert (
-                result.stderr.find(
-                    "Invalid value for '--role': invalid choice: ADMN. (choose from ADMIN)"
-                )
-                > -1
+    def test_user_creation_with_invalid_role(db_session, server_config, pytestconfig):
+        runner = CliRunner(mix_stderr=False)
+        result = runner.invoke(
+            user_create,
+            args=[
+                TestUserManagement.USER_SWITCH,
+                TestUserManagement.USER_TEXT,
+                TestUserManagement.PSWD_SWITCH,
+                TestUserManagement.PSWD_TEXT,
+                TestUserManagement.EMAIL_SWITCH,
+                TestUserManagement.EMAIL_TEXT,
+                TestUserManagement.FIRST_NAME_SWITCH,
+                TestUserManagement.FIRST_NAME_TEXT,
+                TestUserManagement.LAST_NAME_SWITCH,
+                TestUserManagement.LAST_NAME_TEXT,
+                TestUserManagement.ROLE_SWITCH,
+                "ADMN",
+            ],
+        )
+        assert result.exit_code == 2, result.stderr
+        assert (
+            result.stderr.find(
+                "Invalid value for '--role': invalid choice: ADMN. (choose from ADMIN)"
             )
+            > -1
+        )
 
     @staticmethod
-    @responses.activate
-    def test_valid_user_delete(client, server_config, pytestconfig):
-        with client:
-            runner = CliRunner(mix_stderr=False)
-            result = TestUserManagement.register_user()
-            # Make sure register_user helper function worked
-            assert result.stdout_bytes == b"User test_user registered\n"
-            assert result.exit_code == 0, result.stderr
+    def test_valid_user_delete(db_session_shared, pytestconfig):
+        # with db_session_shared:
+        runner = CliRunner(mix_stderr=False)
+        result = TestUserManagement.register_user()
+        # Make sure register_user helper function worked
+        assert result.stdout_bytes == b"User test_user registered\n"
+        assert result.exit_code == 0, result.stderr
 
-            result = runner.invoke(user_delete, args=[TestUserManagement.USER_TEXT])
-            assert result.exit_code == 0, result.stderr
+        result = runner.invoke(user_delete, args=[TestUserManagement.USER_TEXT])
+        assert result.exit_code == 0, result.stderr
```

Comment:
I'm really tempted to say that it'd be cleaner to use mocks on the `User` model operations and run a _single_ command in each test case rather than stringing them together like this; this starts to smell more like the current gold unittest monolith than true "unit" test cases.

That is, in this case, apply mocks to make `user_delete` see a `User` in the proper state rather than trying to actually create one and worrying about DB persistence across the pieces.

Reply:
This gets a little complex, `user_delete` actually creates a standalone db_session that queries the database for delete operation. I am not sure how I can make user_delete see a User without actually creating one. I know we can do something similar by mocking the db query results but not looked into it in detail. Anyway this test uses a db file from our temp directory so it gets cleaned up after every test (very similar to our tests that uses in-memory database).

Reply:
I think Dave has a good point:  once we need a database which persists across tests (even if it's in a temporary file), we've moved from _unit_ testing to _functional_ testing.

The key to unit testing is the ability to define a "unit".  All inputs to the unit are controlled by the test code, and dependencies of the code-under-test (CUT) which are outside the unit should be mocked.  So, in this case, the calls to database methods from the CUT should not actually reach the database code -- the database implementation (I would suggest...) is outside the "unit".  The test should be supplying mocks for the database calls which drive the behavior of the CUT down the intended paths.  Python and PyTest make it _really_ easy to do this sort of mocking.  If you cannot come up with something better `monkeypatch` will let you redirect the database methods to code which the test controls.

Reply:
Understood, I think I have found a way to do that using `monkeypatch.setattr` functionality but I think I will address this in another PR. In my conversation with @portante we have found out couple pieces of code that needs a refactor as well, so I will address this along with the refactor.